File: src/main/java/leetcode/string/onetwofive/ValidPalindrome_125.java
package leetcode.string.onetwofive;

public class ValidPalindrome_125 {



}

class Solution {
    public boolean isPalindrome(String s) {
        String factor = s.toLowerCase();
        int start = 0;
        int end = factor.length() - 1;
        while (start < end) {
            //start 포인터의 char가 영문자나 숫자가 아닐 경우
            if(!Character.isLetterOrDigit(factor.charAt(start))) {
                start++;
            } else if(!Character.isLetterOrDigit(factor.charAt(end))) {
                end--;
            } else {
                if(factor.charAt(start) != factor.charAt(end)){
                    return false;
                }
                start++;
                end--;
            }

        }

        return true;
    }
}

File: src/main/java/leetcode/string/onetwofive/ValidPalindrome_125.kt
package leetcode.string.onetwofive

fun isPalindrome(s: String): Boolean {
    var start = 0;
    var end = s.length - 1

    while(start < end) {
        when {
            !Character.isLetterOrDigit(s[start]) -> start++
            !Character.isLetterOrDigit(s[end]) -> end--
            else -> {
                if(Character.toLowerCase(s[start]) != Character.toLowerCase(s[end])) {
                    return false
                }
                start++
                end--
            }
        }
    }
    return true
}

fun main() {
    var s = "A man, a plan, a canal: Panama"
    if (isPalindrome(s).equals(true)) {
        println("O")
    }
}

